cmake-edit-core/src/main/java/ru/urururu/cmakeedit/core/checker/SimulationState.java

```diff
@@ -12,19 +12,109 @@
 
-    SimulationState(List<CommandInvocationNode> nodes, int position) {
-        this(nodes, position, new HashMap<>());
+    SimulationState(List<CommandInvocationNode> nodes, int position, Set<CommandInvocationNode> suspiciousPoints) {
+        this(nodes, position, suspiciousPoints, new HashMap<>());
     }
 
-    SimulationState(List<CommandInvocationNode> nodes, int position, Map<String, Set<CommandInvocationNode>> variables) {
+    SimulationState(List<CommandInvocationNode> nodes, int position, Set<CommandInvocationNode> suspiciousPoints, Map<String, Set<CommandInvocationNode>> variables) {
         this.nodes = nodes;
         this.position = position;
         this.variables = variables;
+        this.suspiciousPoints = suspiciousPoints;
     }
 
-    void simulate(Set<Node> suspiciousPoints, Node node) {
+    String getNodeFromState(Node argumentNode) {
+        StringBuilder sb = new StringBuilder();
+
+        argumentNode.visitAll(new NodeVisitorAdapter() {
+            @Override
+            public void accept(ExpressionNode node) {
+                StringBuilder exprBuilder = new StringBuilder();
+
+                if (node.getKey() != null) {
+                    exprBuilder.append(node.getKey());
+                }
+
+                node.getNested().stream().forEach(n -> exprBuilder.append(getNodeFromState(n)));
+
+                String expr = exprBuilder.toString();
+                getValue(expr);
+
+                sb.append('*').append(expr);
+            }
+
+            @Override
+            public void accept(ConstantNode node) {
+                sb.append(node.getValue());
+            }
+        });
+
+        return sb.toString();
+    }
+
+    String getValue(ArgumentNode argumentNode) {
+        getNodeFromState(argumentNode);
+
+        return getValue(getArgument(argumentNode));
+    }
+
+    String getValue(String expression) {
+        int expressionStart = expression.indexOf("${"); // todo add support for ENV
+        int expressionEnd = expression.lastIndexOf('}');
+
+        String result;
+        if (expressionStart != -1 && expressionEnd > expressionStart) {
+            String sub = expression.substring(expressionStart + 2, expressionEnd);
+
+            // todo remove suspicious point for sub or for getValue(sub)?
+
+            result = expression.substring(0, expressionStart) + getValue(sub) + expression.substring(expressionEnd);
+        } else {
+            result = expression;
+        }
+
+        suspiciousPoints.removeAll(variables.getOrDefault(result, Collections.emptySet()));
+
+        return result;
+    }
+
+    void putValue(ArgumentNode argumentNode, CommandInvocationNode command) {
+        putValue(argumentNode, command, false);
+    }
+
+    void putValue(ArgumentNode argumentNode, CommandInvocationNode command, boolean parentScope) {
+        // expressions from argumentNode should be used
+
+        // todo inline argumentNode via getValue()
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1135)